icekit/images_api/tests.py

```diff
@@ -0,0 +1,372 @@
+                'image': open(self.image.image.file.name, 'rb'),
+            },
+            HTTP_AUTHORIZATION='Token %s' % self.superuser_token.key,
+        )
+        self.assertEqual(201, response.status_code)
+        new_image = Image.objects.get(pk=response.data['id'])
+        self.assertEqual('New image', new_image.title)
+
+    def test_replace_image_with_put(self):
+        response = self.client.get(
+            reverse('images-detail', args=[self.image.pk]),
+            HTTP_AUTHORIZATION='Token %s' % self.superuser_token.key)
+        self.assertEqual(200, response.status_code)
+
+        image_data = response.data
+        image_data['title'] = 'Updated image'
+        image_data['image'] = open(self.image.image.file.name, 'rb')
+
+        response = self.client.put(
+            reverse('images-detail', args=[self.image.pk]),
+            image_data,
+            HTTP_AUTHORIZATION='Token %s' % self.superuser_token.key,
+        )
+        self.assertEqual(200, response.status_code)
+        updated_image = Image.objects.get(pk=self.image.pk)
+        self.assertEqual('Updated image', updated_image.title)
+
+    def test_delete_image_with_delete(self):
+        response = self.client.delete(
+            reverse('images-detail', args=[self.image.pk]),
+            HTTP_AUTHORIZATION='Token %s' % self.superuser_token.key)
+        self.assertEqual(204, response.status_code)
+        self.assertEqual(0, Image.objects.count())
+
+    def test_django_model_permissons(self):
+        # User without any model permissions cannot do anything
+        response = self.client.get(
+            reverse('images-list'),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key)
+        self.assertEqual(403, response.status_code)
+
+        response = self.client.get(
+            reverse('images-detail', args=[self.image.pk]),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key)
+        self.assertEqual(403, response.status_code)
+
+        response = self.client.post(
+            reverse('images-list'),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key,
+        )
+        self.assertEqual(403, response.status_code)
+
+        response = self.client.put(
+            reverse('images-list'),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key)
+        self.assertEqual(403, response.status_code)
+
+        response = self.client.patch(
+            reverse('images-detail', args=[self.image.pk]),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key)
+        self.assertEqual(403, response.status_code)
+
+        response = self.client.delete(
+            reverse('images-detail', args=[self.image.pk]),
+            HTTP_AUTHORIZATION='Token %s' % self.user_token.key)
+        self.assertEqual(403, response.status_code)
+
+        # User with 'change' model permissions can list and change existing
+        # but not add or delete
+        self.user.user_permissions = [
```

Comment:
Hrm, I think we need a permission to allow read access independently of change access, and independently of possessing a token - some API consumers may need read permission but not write permission. Or else a specific `can access image API` permission that is necessary to access any part of API and used in combination with django's built-in ones.

We may introduce a public-level permission in future. Whatever we decide should work similarly on ICEkit's other APIs, and with a 'public' addition in future.